<?xml version="1.0" ?>
<!--
  Follow actor plugin demo.

  There are 4 actors, each following a diffetent box.
  Two actors use a walking animation, and two use a running animation.
  Two actors are slow (1 m/s) and two actors are fast (4 m/s).

  Move the boxes with the transform controls tool to see the actors follow them.
-->
<sdf version="1.6">
  <world name="control_actor">
    <plugin
      filename="gz-sim-physics-system"
      name="gz::sim::systems::Physics">
    </plugin>
    <plugin
      filename="gz-sim-user-commands-system"
      name="gz::sim::systems::UserCommands">
    </plugin>
    <plugin
      filename="gz-sim-scene-broadcaster-system"
      name="gz::sim::systems::SceneBroadcaster">
    </plugin>

    <light type="directional" name="sun">
      <cast_shadows>true</cast_shadows>
      <pose>0 0 10 0 0 0</pose>
      <diffuse>0.8 0.8 0.8 1</diffuse>
      <specular>0.2 0.2 0.2 1</specular>
      <attenuation>
        <range>1000</range>
        <constant>0.9</constant>
        <linear>0.01</linear>
        <quadratic>0.001</quadratic>
      </attenuation>
      <direction>-0.5 0.1 -0.9</direction>
    </light>

    <model name="ground_plane">
      <static>true</static>
      <link name="link">
        <collision name="collision">
          <geometry>
            <plane>
              <normal>0.0 0.0 1</normal>
            </plane>
          </geometry>
        </collision>
        <visual name="visual">
          <geometry>
            <plane>
              <normal>0.0 0.0 1</normal>
              <size>100 100</size>
            </plane>
          </geometry>
          <material>
            <ambient>0.8 0.8 0.8 1</ambient>
            <diffuse>0.8 0.8 0.8 1</diffuse>
            <specular>0.8 0.8 0.8 1</specular>
          </material>
        </visual>
      </link>
    </model>

    <model name="red_box">
      <pose>1 -2 0.5 0 0 0</pose>
      <link name="box_link">
        <inertial>
          <inertia>
            <ixx>1</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>1</iyy>
            <iyz>0</iyz>
            <izz>1</izz>
          </inertia>
          <mass>1.0</mass>
        </inertial>
        <collision name="box_collision">
          <geometry>
            <box>
              <size>1 1 1</size>
            </box>
          </geometry>
        </collision>

        <visual name="box_visual">
          <geometry>
            <box>
              <size>1 1 1</size>
            </box>
          </geometry>
          <material>
            <ambient>1 0 0 1</ambient>
            <diffuse>1 0 0 1</diffuse>
            <specular>1 0 0 1</specular>
          </material>
        </visual>
      </link>
    </model>

    <include>
      <name>human</name>
      <!-- Root node pose (torso) -->
      <pose>0 -2 1.0 0 0 0</pose>
      <uri>https://fuel.gazebosim.org/1.0/chapulina/models/Walking actor</uri>
      <plugin filename="libControlActor.so"  name="gz::sim::systems::ControlActor" >
        <min_distance>0.03</min_distance>
        <animation_x_vel>4.58837</animation_x_vel>
        <topic_cmd_vel>cmd_vel_actor</topic_cmd_vel>
        <topic_actor_global_pose>human_global_pose</topic_actor_global_pose>
        <topic_actor_local_pose>human_local_pose</topic_actor_local_pose>
        <topic_pose_follow_actor>pose_control_actor</topic_pose_follow_actor>
        <topic_mode_actor>mode_position_actor</topic_mode_actor>
        <publish_actor_position_global>true</publish_actor_position_global>
         <!-- Note: publish_actor_position_local requires a relative model to function correctly -->
        <publish_actor_position_local>true</publish_actor_position_local>  
        <robot_name>red_box</robot_name>
        <publish_actor_tf_global>true</publish_actor_tf_global>
        <publish_actor_tf_local>true</publish_actor_tf_local>
        <frame_id>odom</frame_id>
        <child_frame_id>human_link</child_frame_id>
        <frame_local_id>base_link</frame_local_id>
        <child_frame_local_id>human_local_link</child_frame_local_id>
        <cmd_vel_timeout>0.5</cmd_vel_timeout> 
        <pose_pub_hz>120</pose_pub_hz>
      </plugin>
    </include>

  </world>
</sdf>